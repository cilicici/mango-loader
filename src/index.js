import path from 'path';

const publicFolder = 'mango'

export default function loader(content) {

  const url = getFileName(this)

  const outputPath = path.posix.join(publicFolder, url);

  const publicPath = `__webpack_public_path__ + ${JSON.stringify(outputPath)}`;

  this.emitFile(outputPath, content);

  const esModule = true;

  return `${esModule ? 'export default' : 'module.exports ='} ${publicPath};`;}

export function getFileName(baseObject) {
  let ext = 'js';
  let basename = 'file';
  const parsed = path.parse(baseObject.resourcePath);

    if (parsed.ext) {
      ext = parsed.ext;
    }

    if (parsed.dir) {
      basename = parsed.name;
    }
  return  `${basename}${ext}`
}

export const raw = true;
