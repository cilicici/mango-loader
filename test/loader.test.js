import {
  compile,
  execute,
  getCompiler,
  normalizeErrors,
  readAsset,
} from './helpers';

import { getFileName } from '../src/index'

describe('loader', () => {
  it('should works without options', async () => {
    const compiler = getCompiler('simple.js');
    const stats = await compile(compiler);

    expect(
      execute(readAsset('main.bundle.js', compiler, stats))
    ).toMatchSnapshot('result');
    expect(Object.keys(stats.compilation.assets)).toMatchSnapshot('assets');

    expect(normalizeErrors(stats.compilation.warnings)).toMatchSnapshot(
      'warnings'
    );
    expect(normalizeErrors(stats.compilation.errors)).toMatchSnapshot('errors');
  });

  it('getFileName', () => {
    const fileObjects = [{
      resourcePath: '/node_modules/mango-loader/file.png'
    }, {
      resourcePath: '/node_modules/mango-loader/file'
    }, {
      resourcePath: ''
    }]

    expect(getFileName(fileObjects[0])).toBe('file.png')
    expect(getFileName(fileObjects[1])).toBe('file.bin')
    expect(getFileName(fileObjects[2])).toBe('file.bin')
  })

  it('should work with "ModuleConcatenationPlugin" plugin', async () => {
    const compiler = getCompiler(
      'simple.js',
      {},
      {
        mode: 'production',
        optimization: {
          minimize: false,
        },
      }
    );
    const stats = await compile(compiler);

    expect(
      execute(readAsset('main.bundle.js', compiler, stats))
    ).toMatchSnapshot('result');
    expect(Object.keys(stats.compilation.assets)).toMatchSnapshot('assets');
    expect(normalizeErrors(stats.compilation.warnings)).toMatchSnapshot(
      'warnings'
    );
    expect(normalizeErrors(stats.compilation.errors)).toMatchSnapshot('errors');

    if (stats.compilation.modules.size) {
      expect(stats.compilation.modules.size).toBe(2);
    } else {
      expect(stats.compilation.modules.length).toBe(1);
    }
  });
});
