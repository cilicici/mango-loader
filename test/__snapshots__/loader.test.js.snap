// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loader should work with "ModuleConcatenationPlugin" plugin: assets 1`] = `
Array [
  "mango/file.png",
  "main.bundle.js",
]
`;

exports[`loader should work with "ModuleConcatenationPlugin" plugin: errors 1`] = `Array []`;

exports[`loader should work with "ModuleConcatenationPlugin" plugin: result 1`] = `"mango/file.png"`;

exports[`loader should work with "ModuleConcatenationPlugin" plugin: warnings 1`] = `Array []`;

exports[`loader should works without options: assets 1`] = `
Array [
  "mango/file.png",
  "main.bundle.js",
]
`;

exports[`loader should works without options: errors 1`] = `Array []`;

exports[`loader should works without options: result 1`] = `"mango/file.png"`;

exports[`loader should works without options: warnings 1`] = `Array []`;
